Kernel/API/ProcessExposed.cpp

```diff
@@ -0,0 +1,255 @@
+    if (!buffer.write(data_buffer->data() + offset, nread))
+        return KResult(EFAULT);
+
+    return nread;
+}
+
+KResult ProcFSGlobalInformation::refresh_data(FileDescription& description) const
+{
+    ScopedSpinLock lock(m_refresh_lock);
+    auto& cached_data = description.data();
+    if (!cached_data)
+        cached_data = adopt_own_if_nonnull(new ProcFSInodeData);
+    VERIFY(description.data());
+    auto& buffer = static_cast<ProcFSInodeData&>(*cached_data).buffer;
+    if (buffer) {
+        // If we're reusing the buffer, reset the size to 0 first. This
+        // ensures we don't accidentally leak previously written data.
+        buffer->set_size(0);
+    }
+    KBufferBuilder builder(buffer, true);
+    if (!const_cast<ProcFSGlobalInformation&>(*this).output(builder))
+        return ENOENT;
+    // We don't use builder.build() here, which would steal our buffer
+    // and turn it into an OwnPtr. Instead, just flush to the buffer so
+    // that we can read all the data that was written.
+    if (!builder.flush())
+        return ENOMEM;
+    if (!buffer)
+        return ENOMEM;
+    return KSuccess;
+}
+
+KResultOr<size_t> ProcFSProcessInformation::read_bytes(off_t offset, size_t count, UserOrKernelBuffer& buffer, FileDescription* description) const
+{
+    dbgln_if(PROCFS_DEBUG, "ProcFSProcessInformation @ {}: read_bytes offset: {} count: {}", name(), offset, count);
+
+    VERIFY(offset >= 0);
+    VERIFY(buffer.user_or_kernel_ptr());
+
+    if (!description)
+        return KResult(EIO);
+    if (!description->data()) {
+        dbgln("ProcFSGlobalInformation: Do not have cached data!");
+        return KResult(EIO);
+    }
+
+    // Be sure to keep a reference to data_buffer while we use it!
+    RefPtr<KBufferImpl> data_buffer = static_cast<ProcFSInodeData&>(*description->data()).buffer;
+
+    if (!data_buffer || (size_t)offset >= data_buffer->size())
+        return 0;
+
+    ssize_t nread = min(static_cast<off_t>(data_buffer->size() - offset), static_cast<off_t>(count));
+    if (!buffer.write(data_buffer->data() + offset, nread))
+        return KResult(EFAULT);
+
+    return nread;
+}
+
+KResult ProcFSProcessInformation::refresh_data(FileDescription& description) const
+{
+    // For process-specific inodes, hold the process's ptrace lock across refresh
+    // and refuse to load data if the process is not dumpable.
+    // Without this, files opened before a process went non-dumpable could still be used for dumping.
+    auto process = const_cast<ProcFSProcessInformation&>(*this).m_parent_folder->m_associated_process;
+    process->ptrace_lock().lock();
+    if (!process->is_dumpable()) {
+        process->ptrace_lock().unlock();
+        return EPERM;
+    }
+    ScopeGuard guard = [&] {
+        process->ptrace_lock().unlock();
+    };
+    ScopedSpinLock lock(m_refresh_lock);
+    auto& cached_data = description.data();
+    if (!cached_data)
+        cached_data = adopt_own_if_nonnull(new ProcFSInodeData);
```

Comment:
This should be changed to `new (nothrow) ProcFSInodeData`, otherwise the Kernel will assert on the allocation succeeding.

You can also use `try_make<T>` for creating `OwnPtr<T>` and `try_create<T>` for creating `RefPtr<T>`.

See  a391175 for the explanation